poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+	}
+	set := anon.Set{p.insurers[i]}
+	_, err := anon.Verify(sig.suite, msg, set, nil, sig.signature)
+	return err
+}
+
+/* A public wrapper function for sign, produces a signature for a given insurer.
+ * Insurers should use this function to produce a PromiseSignature to express
+ * approval of a Promise.
+ *
+ * Arguments
+ *    i         = the index of the insurer's share
+ *    gKeyPair  = the long term public/private keypair of the insurer.
+ *
+ * Return
+ *   A PromiseSignature object with the signature.
+ *
+ * Note
+ *   It is wise to first call VerifyShare to verify the insurer's share before
+ *   signing on the promise.
+ */
+func (p *Promise) Sign(i int, gKeyPair *config.KeyPair) *PromiseSignature {
+	return p.sign(i, gKeyPair, sigMsg)
+}
+
+/* Verifies a signature from a given insurer
+ *
+ * Arguments
+ *    i   = the index of the insurer in the insurers list
+ *    sig = the PromiseSignature object containing the signature
+ *
+ * Return
+ *   an error if the promise is malformed, nil otherwise.
+ */
+func (p *Promise) VerifySignature(i int, sig *PromiseSignature) error {
+	return p.verifySignature(i, sig, sigMsg)
+}
+
+/* Reveals the secret share that the insurer has been protecting. The insurer
+ * should call this function on behalf of a client after verifying that the
+ * Promiser is non-responsive.
+ *
+ * Arguments
+ *    i        = the index of the insurer
+ *    gkeyPair = the long-term keypair of the insurer
+ *
+ * Return
+ *   the revealed private share
+ */
+func (p *Promise) RevealShare(i int, gKeyPair *config.KeyPair) abstract.Secret {
+	diffieBase := p.keySuite.Point().Mul(p.pubKey, gKeyPair.Secret)
+	share := p.diffieHellmanDecrypt(p.secrets[i], diffieBase)
+	return share
+}
+
+/* Verify that a revealed share is properly formed. This should be called by
+ * clients or others who request an insurer to reveal its shared secret.
+ *
+ * Arguments
+ *    i     = the index of the share
+ *    share = the share to validate.
+ *
+ * Return
+ *   Whether the secret is valid
+ */
+func (p *Promise) VerifyRevealedShare(i int, share abstract.Secret) error {
+	if i < 0 || i >= p.n {
+		return errors.New("Invalid index. Expected 0 <= i < n")
+	}
+	if !p.pubPoly.Check(i, share) {
+		return errors.New("The share failed the public polynomial check.")
+	}
+	return nil
+}
+
+/* Create a proof that the promiser maliciously constructed a shared secret. An
+ * insurer should call this if VerifyShare fails due to the public polynomial
+ * check failing. If it failed for other reasons (such as a bad index) it is not
+ * advised to call this function.
+ *
+ * Arguments
+ *    i         = the index of the malicious shared secret
+ *    gKeyPair  = the long term key pair of the insurer of share i
+ *
+ * Return
+ *   A BlameProof that the promiser is malicious or nil if an error occurs
+ *   An error object denoting the status of the proof construction
+ */
+func (p *Promise) Blame(i int, gKeyPair *config.KeyPair) (*BlameProof, error) {
```

Comment:
Similarly, there should be no need for public Blame or VerifyBlameProof methods - they should probably be private.  The only public method should produce a "PromiseResponse" struct or some such that can be either a signature (if a received-and-unmarshalled promise was valid) or a blame proof (if it wasn't).